### PR TITLE
Hyperparameter tuning

### LINKED ISSUE #40: Implement train/test/validation strategy and hyper-parameter tuning.
We want to decide which data to use for tasks of train, test and validation. One approach is mentioned in this [comment](https://github.com/ESMValGroup/ClimaNet/pull/37#issue-4254539417) which is selecting different years in the dataset for each task. 

 Also, we should come up with a best list of values for model parameters. how to do hyper parameter tuning?

### PR DESCRIPTION
**Pull request by SarahAlidoost**
closes #40

### CODE REVIEW

**Comment by SarahAlidoost** [comment]
@meiertgrootes and @rogerkuou this PR adds support for hyper parameter tuning using ray. I added an example notebook that can be run locally using cpu. In meantime, I'm testing the workflow in levante. See #40 for details on tuning strategy.

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost for now I will only review and test this on my local. Will wait for your signal about the HPC test

File: climanet/utils.py [hunk 1]
```diff
@@ -468,6 +489,40 @@ def plot_histograms(
     plt.show()
 
 
+def data_split(
+    data_folder,
+    filename_pattern="*_hr_ERA5dc_masked_tos.nc",
+    train_range=(2018, 2020),
+    validation_range=(2021, 2021),
+    test_range=(2022, 2022),
```
Comment (rogerkuou) [inline climanet/utils.py:498]: Just a small opinion here: maybe these default arguments are too strong constraints.

File: notebooks/example_tuning.ipynb
Comment (rogerkuou) [inline notebooks/example_tuning.ipynb:1]: Comment for the last cell, the `experiment_path` changes per run. I propose it can be changed to:

```python
check_best_model(Path(best_result.path).parent, test_dataset, run_dir)
```

so two changes:
1. I think `check_best_model` can accept `pathlib.Path` so no need to convert to str?
2. Use `Path(best_result.path).parent` torefer the experiment path

File: climanet/tune.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            max_concurrent_trials=tune_config["max_concurrent_trials"],
+        ),
+        param_space=tune_config,
+        run_config=ray.tune.RunConfig(storage_path=tune_config["run_dir"]),
+    )
+    results = tuner.fit()
+    return results
+
+
+def check_best_model(experiment_path: str, test_dataset: STDataset, run_dir):
```
Comment (rogerkuou) [inline climanet/tune.py]: I think both paths here can also accept `pathlib.Path`?

```suggestion
def check_best_model(experiment_path: [str | Path], test_dataset: STDataset, run_dir: [str | Path]):
```

File: climanet/tune.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        verbose=False,
+        tune_checkpoint=True,
+    )
+
+
+def tune_data_preparation(data_config: dict, is_hourly=True) -> STDataset:
+    """Prepare the data for training and validation."""
+    input_data = xr.open_mfdataset(data_config["input_filenames"])
+    monthly_data = xr.open_mfdataset(data_config["monthly_filenames"])
+    lsm_mask = xr.open_dataset(data_config["landmask_filename"])
```
Comment (rogerkuou) [inline climanet/tune.py]: I have some small concerns of wrapping this into a function, since this hides API of `xr.open_mfdataset`

With our current data, for Python<=3.13 this works. While I am using Python=3.14, where the CF concention changed. Then loading `input_data` and `monthly_data` will cause kernel error if not specifying chunksizes in `open_mfdataset`. But this also relates to the Python version where nc files are created.

Maybe we should either expose the args of open_mfdataset, or mention Python<=3.13 in `pyproject.toml` ?

**Review by rogerkuou — APPROVED** [review]
Hi @SarahAlidoost , thanks! Just some small comments. After fixing them please feel free to merge!

File: climanet/utils.py (hunk 1 above)
Comment (SarahAlidoost) [reply climanet/utils.py:498]: The defaults show our strategy for train/test/validation explained [here](https://github.com/ESMValGroup/ClimaNet/issues/40#issuecomment-4877040064). I couldn't understand what is meant by "_too strong constraints_"? Could you explain what constraints are implied by those default values? any suggestions how to fix it?

File: notebooks/example_tuning.ipynb
Comment (SarahAlidoost) [reply notebooks/example_tuning.ipynb:1]: > Comment for the last cell, the `experiment_path` changes per run. I propose it can be changed to:
> 
> ```python
> check_best_model(Path(best_result.path).parent, test_dataset, run_dir)
> ```
> 
> so two changes:
> 
> 1. I think `check_best_model` can accept `pathlib.Path` so no need to convert to str?

right! I checked it, it accepts both path and str. it is fixed now. 

> 2. Use `Path(best_result.path).parent` torefer the experiment path

This is one way of getting the experiment path. If the `best_result` is not available in the current run, the path should be defined by user. I added some comments to the notebook.

File: climanet/tune.py (hunk 2 above)
Comment (SarahAlidoost) [reply climanet/tune.py]: fixed!

File: climanet/tune.py (hunk 3 above)
Comment (SarahAlidoost) [reply climanet/tune.py]: > I have some small concerns of wrapping this into a function, since this hides API of `xr.open_mfdataset`

I agree with this and I think `chunks` argument is useful but only if we refactor `STDataset` class. I added `chunks` to the `xr.open_mfdataset` in `tune_data_preparation`, but with all `.values` and `from_numpy` calls in  `STDataset` class, data is not lazy and the benefit of chunking is mostly lost. I submitted issue #75 to refactor `STDataset` class to keep data lazy.   

> With our current data, for Python<=3.13 this works. While I am using Python=3.14, where the CF concention changed. Then loading `input_data` and `monthly_data` will cause kernel error if not specifying chunksizes in `open_mfdataset`. But this also relates to the Python version where nc files are created.
> 

I can call `xr.open_mfdataset` with Python=3.14 without specifying chunk. Also, in xarray documentation, `chunks` is still an optional argument. Could you add a reference/link for this problem related to "CF", "python version" and "chunksize"? 

> Maybe we should either expose the args of open_mfdataset, or mention Python<=3.13 in `pyproject.toml` ?

**Comment by SarahAlidoost** [comment]
> Hi @SarahAlidoost , thanks! Just some small comments. After fixing them please feel free to merge!

@rogerkuou Hi, thanks for the review. I won't merge this PR before getting results of the notebook on levante, checking ray and gpus.

File: climanet/tune.py (hunk 3 above)
Comment (rogerkuou) [reply climanet/tune.py]: Okay, I think it's related to: https://github.com/pydata/xarray/issues/11205

In summary, this is likely to be an numpy bug for Python3.14 and only occurs to WSL. I will use 3.13 myself. We do not need to further look into this then

File: notebooks/example_tuning.ipynb
Comment (rogerkuou) [reply notebooks/example_tuning.ipynb:1]: thanks @SarahAlidoost ! Just one more thing, in your current version, I think `results.experiment_path` points to one level deeper: `......./runs_daily/_train_2026-07-16_16-48-59/_train_7a5a1_00001_1_patch_size=4_2026-07-16_16-48-59`, instead of ` `......./runs_daily/_train_2026-07-16_16-48-59/`. Hence I added the `.parent`, maybe we should still keep it?

I agree the comment is already quite useful though.

File: climanet/utils.py (hunk 1 above)
Comment (rogerkuou) [reply climanet/utils.py:498]: Sorry I was not clear. I meant the range are very specific to this dataset, and may hide the configuration API if someone wants to use other years for train/validation/test. I am more in favor of keep them as necessary input arguement, but only keep the default values in the example notebook/scripts.

But this is really a minor thing so feel free to also keep it as it is.

File: climanet/utils.py [hunk 4]
```diff
@@ -468,6 +489,40 @@ def plot_histograms(
     plt.show()
 
 
+def data_split(
```
Comment (meiertgrootes) [inline climanet/utils.py:493]: Two (semi-)minor points here:

1) this assumes that train/test/validation will cover contiguous ranges in time. That is a choice we are making, but may be sth that would benefit from being more flexible. For the time being I'd say leave it as it is, in order to proceed, but maybe worth returning to as a separate issue

2) to nit pick a little bit: technically hyper-parameter tuning, which is performed on the basis of the test score results requires an additional hold-out set to finally asses model performance on pristine data. Since we are only extending to 2022 we could use 2023, but this does open the issue of actual physical temporal drift.

File: climanet/tune.py [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+            },
+        ),
+        tune_config=ray.tune.TuneConfig(
+            metric="loss",
+            mode="min",
+            scheduler=scheduler,
+            num_samples=tune_config["num_trials"],
+            max_concurrent_trials=tune_config["max_concurrent_trials"],
+        ),
+        param_space=tune_config,
```
Comment (meiertgrootes) [inline climanet/tune.py]: this will feed all entries in tune_config in as parameters in the Hyperparameter optimization. This will (likely) result in things like num_trials, cpu+per_trial etc. being tracked in logs and dashboards as constant parameters, impacting overview. Could we consider cleaning out non-optimisation parameters, before injection?

**Review by meiertgrootes — APPROVED** [review]
Some minor comments, but nothings to stop merging.

File: climanet/tune.py (hunk 5 above)
Comment (SarahAlidoost) [reply climanet/tune.py]: good point! In [their example](https://docs.ray.io/en/latest/_collections/tune/examples/tune_pytorch_asha/tune_pytorch_asha.html#define-the-hyperparameter-search-space-and-configuration), both static arguments and hyper parameters are in one config dictionary passed to the training function. But as you mentioned, it is cleaner to split them. I found that is possible to pass arguments in `ray.tune.with_parameters` and fixed this.

File: climanet/utils.py (hunk 4 above)
Comment (SarahAlidoost) [reply climanet/utils.py:493]: 1. agree, I've explained the choice [here](https://github.com/ESMValGroup/ClimaNet/issues/40#issuecomment-4877040064). In the [MAESSTRO code](https://github.com/jpl-slice/maesstro/blob/98dab19479ef86d289fa3e9607c5ebe405f39be5/main_pretrain.py#L110), they did similar, different years for training and validation (like 2012 vs 2011). I'll submit a new issue to consider random split in time, which needs trying out different strategies again. 
2. Right, we have three datasets: train/validation/test. In our training workflow, we use validation set and hold out the test data. The model haven't seen the test data.

File: notebooks/example_tuning.ipynb
Comment (SarahAlidoost) [reply notebooks/example_tuning.ipynb:1]: in my version `results.experiment_path` points to first level e.g. `./runs_daily/_train_2026-07-16_16-48-59/`. Can you check your version again? Note in the last commit I removed the function `check_best_model` and moved everything to the notebook.

**Comment by SarahAlidoost** [comment]
> Some minor comments, but nothings to stop merging.

@meiertgrootes thanks for the review. I addressed your comments. I'm still testing this PR on HPC using cuda. Once passed, I'll merge this PR.

**Comment by SarahAlidoost** [comment]
@meiertgrootes and @rogerkuou I was able to set up Ray on levante with several nodes and started a test tune run with 20 nodes. The settings and scripts are in the [`scripts` folder](https://github.com/ESMValGroup/ClimaNet/tree/main/scripts), see `tuning.slurm` and `tuning.py`. I also add a documentation explaining the choices, see [`docs/tune.md`](https://github.com/ESMValGroup/ClimaNet/blob/main/docs/tune.md). The **main** change is that I could not use 3 years of data as train_set but only 1 year. Because the STDataset gets memory error, this will be fixed in #75. After tuning, the model is usually re-trained anyway. So then we can use 3 years of data. 

for now, I'll merge this PR, if the test runs fails due to memory error or timeout, I'll adjust the sbatch settings and later I'll push changes in a new PR.
